File: src/workbook-export.ts
import ExcelJS from "exceljs";
import type { Dict } from "./db.ts";

export type Table = {
  title: string;
  source?: string;
  columns: [string, string][];
  rows: Dict[];
  notes?: string[];
};

export async function workbookBuffer(tables: Table[]) {
  const wb = new ExcelJS.Workbook();
  wb.creator = "Quản lý thi đua THPT Giao Thủy C";
  wb.subject = "Báo cáo thi đua";
  const used = new Set<string>();
  for (const table of tables) {
    let name = table.title.replace(/[\\/*?:\[\]]/g, "-").slice(0, 31) || "Bang";
    let suffix = 2;
    while (used.has(name)) {
      const tag = `-${suffix++}`;
      name = `${name.slice(0, Math.max(1, 31 - tag.length))}${tag}`;
    }
    used.add(name);
    const ws = wb.addWorksheet(name);
    const cols = table.columns;
    const width = Math.max(1, cols.length);
    ws.mergeCells(1, 1, 1, width);
    ws.getCell(1, 1).value = table.title;
    ws.getCell(1, 1).font = { size: 15, bold: true, color: { argb: "FF17365D" } };
    const notes = [table.source ?? "", ...(table.notes ?? [])].filter(Boolean);
    notes.forEach((note, i) => {
      const row = i + 2;
      ws.mergeCells(row, 1, row, width);
      ws.getCell(row, 1).value = note;
      ws.getCell(row, 1).alignment = { wrapText: true, vertical: "top" };
    });
    const headerRow = notes.length + 2;
    cols.forEach(([_, label], i) => {
      const cell = ws.getCell(headerRow, i + 1);
      cell.value = label;
      cell.font = { bold: true, color: { argb: "FFFFFFFF" } };
      cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FF17365D" } };
      cell.border = { top: { style: "thin" }, left: { style: "thin" }, bottom: { style: "thin" }, right: { style: "thin" } };
      ws.getColumn(i + 1).width = Math.min(35, Math.max(13, label.length + 2));
      cell.alignment = { vertical: "middle", horizontal: "center", wrapText: true };
    });
    table.rows.forEach((record, r) => {
      const rowIdx = headerRow + 1 + r;
      cols.forEach(([key], c) => {
        const cell = ws.getCell(rowIdx, c + 1);
        const v = record[key];
        if (typeof v === "string" && /^[=+\-@]/.test(v)) {
          cell.value = v;
          cell.numFmt = "@";
        } else {
          cell.value = v == null || Array.isArray(v) || typeof v === "object" ? null : (v as string | number | boolean);
        }
        cell.border = { top: { style: "thin" }, left: { style: "thin" }, bottom: { style: "thin" }, right: { style: "thin" } };
        if ((r + 1) % 2 === 0) cell.fill = { type: "pattern", pattern: "solid", fgColor: { argb: "FFEDF3F8" } };
        cell.alignment = { vertical: "top", wrapText: true };
        if (typeof v === "number" && !Number.isInteger(v)) cell.numFmt = "0.0000";
      });
    });
    const last = headerRow + Math.max(1, table.rows.length);
    ws.autoFilter = { from: { row: headerRow, column: 1 }, to: { row: headerRow, column: width } };
    ws.views = [{ state: "frozen", xSplit: Math.min(2, width), ySplit: headerRow }];
    ws.pageSetup = {
      orientation: width > 10 ? "landscape" : "portrait",
      fitToPage: true, fitToWidth: 1, fitToHeight: 0, paperSize: 9,
      printArea: `A1:${ws.getColumn(width).letter}${last}`,
      printTitlesRow: `${headerRow}:${headerRow}`,
    };
    ws.headerFooter.oddFooter = "Trang &P / &N";
  }
  return Buffer.from(await wb.xlsx.writeBuffer());
}

File: quanlythidua/workbook_export.py
"""Xuất các bảng nghiệp vụ ra Excel, giữ nguyên số và ô chưa có dữ liệu."""
from __future__ import annotations

from io import BytesIO
import re

from flask import send_file
from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill
from openpyxl.utils import get_column_letter


def _text(cell, value):
    """Nội dung người dùng là văn bản, không phải công thức Excel."""
    cell.value = value
    if isinstance(value, str):
        cell.data_type = "s"


def workbook_bytes(tables):
    wb = Workbook()
    wb.remove(wb.active)
    for table in tables:
        title = table["title"]
        sheet_name = re.sub(r"[\\/*?:\[\]]", "-", title)[:31] or "Bảng"
        ws = wb.create_sheet(sheet_name)
        cols = table["columns"]
        width = max(1, len(cols))
        _text(ws.cell(1, 1), title)
        ws.merge_cells(start_row=1, start_column=1, end_row=1, end_column=width)
        ws.cell(1, 1).font = Font(size=15, bold=True, color="17365D")
        notes = [table.get("source", ""), *table.get("notes", [])]
        notes = [note for note in notes if note]
        for row, note in enumerate(notes, 2):
            _text(ws.cell(row, 1), note)
            ws.merge_cells(start_row=row, start_column=1, end_row=row, end_column=width)
            ws.cell(row, 1).alignment = Alignment(wrap_text=True, vertical="top")
            ws.row_dimensions[row].height = 32
        header_row = len(notes) + 2
        for col, (key, label) in enumerate(cols, 1):
            cell = ws.cell(header_row, col)
            _text(cell, label)
            cell.font = Font(bold=True, color="FFFFFF")
            cell.fill = PatternFill("solid", fgColor="17365D")
            cell.alignment = Alignment(wrap_text=True, vertical="center")
            ws.column_dimensions[get_column_letter(col)].width = min(35, max(13, len(label) + 2))
        ws.row_dimensions[header_row].height = 40
        for row_idx, record in enumerate(table["rows"], header_row + 1):
            for col, (key, _) in enumerate(cols, 1):
                cell = ws.cell(row_idx, col)
                _text(cell, record.get(key))
                cell.alignment = Alignment(vertical="top", wrap_text=True)
                if isinstance(cell.value, (int, float)):
                    cell.number_format = "0.##########"
                if (row_idx - header_row) % 2 == 0:
                    cell.fill = PatternFill("solid", fgColor="EDF3F8")
        ws.freeze_panes = f"C{header_row + 1}"
        ws.auto_filter.ref = f"A{header_row}:{get_column_letter(width)}{max(header_row, ws.max_row)}"
        ws.print_title_rows = f"1:{header_row}"
        ws.sheet_properties.pageSetUpPr.fitToPage = True
        ws.page_setup.orientation = "landscape"
        ws.page_setup.paperSize = ws.PAPERSIZE_A3 if width > 15 else ws.PAPERSIZE_A4
        ws.page_setup.fitToWidth = 1
        ws.page_setup.fitToHeight = 0
    if not wb.sheetnames:
        raise ValueError("Không có bảng để xuất")
    stream = BytesIO()
    wb.save(stream)
    stream.seek(0)
    return stream


def send_tables(tables, filename):
    return send_file(
        workbook_bytes(tables),
        as_attachment=True,
        download_name=filename,
        mimetype="application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
    )
